File: Módulo 3/coletaDadosMySQL.r
#Instalação do pacote
install.packages('RMariaDB')

#importação do pacote

#Importar pacote RMariaDB se ele ainda não foi carregado
if (!"RMariaDB" %in% (.packages())){require(RMariaDB)}

#Conectar ao SGDB MySQL --> Banco de Dados bootcamp
con <- dbConnect(MariaDB(), user = "root", password = "root",
                dbname = "bootcamp", host = "localhost", serverTimezone = 'UTC')

#Para listar quais tabelas existem no esquema .bootcamp execute:

#Ler a lista de tabelas no DB
tables <- dbListTables(con)
tables

#Para consultar os dados de uma tabela, execute:
#dbReadTable(nome-da-conexao, "nome-da-tabela")

#Consultar os dados da tabela *estado*
tabledata <- dbReadTable(con, "estado")
tabledata

#Consultar os dados da tabela *cidade*
tabledata <- dbReadTable(con, "cidade")
tabledata

#Consultar os dados da tabela *imovel*
tabledata <- dbReadTable(con, "imovel")
tabledata

#Consultar os dados da tabela *tipounidade*
tabledata <- dbReadTable(con, "tipounidade")
tabledata

#Para executar um comando SQL execute:
#dbSendQuery(nome-da-conexão, "comando")

#Vamos inserir uma nova linha na tabela tipounidade
#Criar o comando e salvar na variável query
query <- "INSERT INTO tipounidade(idTipoUnidade, dscTipoUnidade) VALUES (7, 'Loft');"

#Opcional
print(query)

results <- dbSendQuery(con, query)
results

#Limpar resultados
dbClearResult(results)

#Consultar os dados da tabela *tipounidade*
tabledata <- dbReadTable(con, "tipounidade")
tabledata

#Inserção por parâmetros
id <- 8
desc <- 'Chácara'

query <- paste("INSERT INTO tipounidade(idTipoUnidade, dscTipoUnidade) VALUES (",id,",'",desc,"');", sep = '')

#Opcional
print(query)

results <- dbSendQuery(con, query)
results

#Limpar resultados
dbClearResult(results)

#Consultar os dados da tabela *tipounidade*
tabledata <- dbReadTable(con, "tipounidade")
tabledata

#Instalação de pacote
install.packages('xlsx')

#importação do pacote

#Importar pacote xlsx se ele ainda não foi carregado
if (!"xlsx" %in% (.packages())){require(xlsx)}

datasetpath <- 'C:/Users/luciano/IGTI/dados'

filename <- paste(datasetpath, "/estados.xlsx", sep = '')
filename

insertdata <- read.xlsx(filename, sheetIndex = 1, header = TRUE, encoding = "UTF-8")
insertdata

dbWriteTable(con, 'estado', insertdata, append = TRUE)

#Consultar os dados da tabela *estado*
tabledata <- dbReadTable(con, "estado")
tabledata

#Importar pacote xlsx se ele ainda não foi carregado
if (!"xlsx" %in% (.packages())){require(xlsx)}

query <- "SELECT * FROM estado;"

results <- dbGetQuery(con, query)
results

#Salva o resultado da query no arquivo CSV
write.csv(results, "C:/Users/luciano/IGTI/dados/estadosDB.csv", row.names = FALSE, quote = FALSE)

#dbCommit(con)

#Criar uma tabela vazia
query <- "CREATE TABLE caracteristicasgerais(
                idcaracteristicasGerais int NOT NULL AUTO_INCREMENT,
                dsccaracteristicasGerais varchar(150) NOT NULL,
                PRIMARY KEY (idcaracteristicasGerais));"

results <- dbSendQuery(con, query)
results

tables <- dbListTables(con)
tables

#Desconectar o Banco de Dados
dbDisconnect(con)

File: Módulo 3/coletaDadosMySQL.py
#!/usr/bin/env python
# coding: utf-8

# # Coleta de Dados no MySQL utilizando o Python

# ### Para mais detalhes sobre o pacote mysql.connector consulte:

# * https://www.w3schools.com/python/python_mysql_getstarted.asp


#Importar pacote de conexão mysql
import mysql.connector


# Para instalação do pacote, execute o comando abaixo no prompt do Anaconda:
# 
# * conda install -c anaconda mysql-connector-python
# 

# #### Conectar ao MySQL Server sem selecionar um banco de dados

mydb = mysql.connector.connect(
    host = 'localhost',
    user = 'root',
    password = 'root'
)

print(mydb)

mycursor = mydb.cursor()

mycursor.execute("SHOW DATABASES")

for db in mycursor:
    print(db)


# #### Para criar um banco de dados use o comando abaixo

mycursor.execute("CREATE DATABASE mydatabase")


# #### Conectar ao MySQL Server selecionando um banco de dados

mydb = mysql.connector.connect(
    host = 'localhost',
    user = 'root',
    password = 'root',
    database = 'bootcamp'
)

print(mydb)

mycursor = mydb.cursor()


# #### Exibir tabelas na base de dados

mycursor.execute("SHOW TABLES")

for db in mycursor:
    print(db)


# #### Criar uma tabela

mycursor.execute("CREATE TABLE caracteristicageralimovel (    idCaracteristicasImovel int NOT NULL AUTO_INCREMENT,     idcaracteriticasGerais int NOT NULL,     idImovel int NOT NULL,     PRIMARY KEY (idCaracteristicasImovel),     KEY fk_caracteriticasGerais_has_Imovel_Imovel1_idx (idImovel),     KEY fk_caracteriticasGerais_has_Imovel_caracteristicasGerais1_idx (idcaracteriticasGerais),     CONSTRAINT fk_caracteriticasGerais_has_Imovel_caracteristicasGerais1 FOREIGN KEY (idcaracteriticasGerais)                    REFERENCES caracteristicasgerais (idcaracteristicasGerais),    CONSTRAINT fk_caracteriticasGerais_has_Imovel_Imovel1_idx FOREIGN KEY (idImovel)                    REFERENCES imovel (idImovel))")


# #### Verificar tabela construída

mycursor.execute("SHOW TABLES")

for db in mycursor:
    print(db)


# #### Inserir registro em tabela

query = "INSERT INTO caracteristicasgerais (idcaracteristicasGerais, dsccaracteristicasGerais) VALUES (%s, %s)"
values = (1, "Portaria 24 horas")
mycursor.execute(query, values)

mydb.commit()

print(mycursor.rowcount, "registro(s) inserido(s).")


# #### Inserindo múltiplos valores

values = [(2, "Elevador"), (3, "Piscina")]
mycursor.executemany(query, values)

mydb.commit()

print(mycursor.rowcount, "registro(s) inserido(s).")


# #### Consultar toda a tabela

mycursor.execute("SELECT * FROM caracteristicasgerais")

myresult = mycursor.fetchall()

for mydata in myresult:
    print(mydata)


# #### Consultar um registro

mycursor.execute("SELECT idcaracteristicasGerais, dsccaracteristicasGerais FROM caracteristicasgerais")

myresult = mycursor.fetchone()

print(myresult)


# #### Importação de arquivo CSV

import csv

#Leitura de arquivo sem utilizar biblioteca Pandas

filename = 'C:/Users/luciano/IGTI/dados/cidades.csv'

with open (filename, 'r') as incsvfile:
    reader = csv.reader(incsvfile, delimiter = ',')
    next(reader, None) #skip the headers
    for line in reader:
        #print(line)
        #print(line[1])
        mycursor.execute ("INSERT INTO cidade (CodigoCompletoIBGE, CodigoCidadeIBGE, NomeCidade, CodEstadoIBGE)                     VALUES (%s, %s, %s, %s)", line)

mydb.commit()

print(mycursor.rowcount, "registro(s) inserido(s).")

incsvfile.close()


# #### Consultar quantidade de itens inseridos

mycursor.execute("SELECT count(*) FROM cidade")

myresult = mycursor.fetchone()

print(myresult)


# #### Importação de arquivo CSV (outra forma)

import csv

#Leitura de arquivo
csv_data = csv.reader('C:/Users/luciano/IGTI/dados/cidades.csv')
csv_data

for row in csv_data:
    i = 0
    print(row)
    if i > 3:
        break
        
#for row in csv_data:
#    mycursor.execute ("INSERT INTO cidade (CodigoCompletoIBGE, CodigoCidadeIBGE, NomeCidade, CodEstadoIBGE) \
#                    VALUES (%s, %s, %s, %s)", row)
